### ISSUE TITLE
Issues due to OCaml 5 compiler bug with `Atomic.get`

### ISSUE THREAD

**Issue by polytypic** [ocaml-multicore/saturn#97]
Today we identified a compiler bug in OCaml 5 where the compiler may incorrectly optimize away repeated `Atomic.get`s.

We already identified several places in my PR #83 that were affected by this compiler bug.

I'll go through the code in Saturn (in the following days) and try to identify other places affected by the bug and add them as comments here.  I already identified one case, but I probably don't have time today to find more.

**Comment by polytypic** [comment]
The MPMC relaxed queue has this operation: https://github.com/ocaml-multicore/saturn/blob/49e4e596460ba24d8fa9059b59a9c784e4414e7f/src_lockfree/mpmc_relaxed_queue.ml#L113-L115

And it is used here: https://github.com/ocaml-multicore/saturn/blob/49e4e596460ba24d8fa9059b59a9c784e4414e7f/src_lockfree/mpmc_relaxed_queue.ml#L125-L134

Unfortunately the compiler optimizes out the `Atomic.get`.  This means that the compiler defeats the intention of `ccas` to avoid writing to memory and reduce contention.  You can observe that the compiler really optimizes the `Atomic.get` out by examining [the compiler output](https://ocaml.godbolt.org/z/oWEK3jeox).  Specifically, consider the snippet:

```
.L135:
        movq    (%rdi), %rbx
.L131:
        testb   $1, %bl
        jne     .L133
.L134:
        movq    (%rdi), %rax
        cmpq    %rbx, %rax
        jne     .L133
```

The register `%rbx` has the value read from an atomic.  That value is stored into a `ref` cell (on first line).  Then the last three lines first reads the value from the `ref` cell back and compares it against the value in register `%rbx`.

So, the optimization done by the compiler defeats the optimization done by the programmer.

**Comment by polytypic** [comment]
I think that [the compiler bug](https://github.com/ocaml/ocaml/issues/12713) affects more things than I initially thought.

Basically, if you have a non-recursive function (i.e. an inlineable function) that uses `Atomic.get`, then such an operation is no longer necessarily [linearizable](https://cs.brown.edu/~mph/HerlihyW90/p463-herlihy.pdf).

You would expect two linearizable operations on a single thread to have a timeline like:

```
<---x--->  <---y--->
|   |   |
|   |   +- Operation ends
|   +- Operation takes effect
+- Operation starts
```

But with the optimization that is no longer the case.  The timelines are merged and the points (`x` and `y`) at which the operations take effect become the one and same point in time.

This can even happen across operations.  You might have three operations (on a single thread) and then operation in the middle might actually be determined after the last operation as the first and last operation are merged.

So, basically, any operations in Saturn that are non-recursive and just use `Atomic.get` are no longer linearizable.
